@navikt/core/react/src/dialog/root/DialogRoot.tsx

```diff
@@ -0,0 +1,173 @@
+
+interface DialogProps {
+  children: React.ReactNode;
+  /**
+   * Whether the dialog is currently open.
+   */
+  open?: boolean;
+  /**
+   * Whether the dialog is initially open.
+   *
+   * To render a controlled dialog, use the `open` prop instead.
+   * @default false
+   */
+  defaultOpen?: boolean;
+  /**
+   * Event handler called when the dialog is opened or closed.
+   */
+  onOpenChange?: (open: boolean, event?: Event) => void;
+  /**
+   * Event handler called after any animations complete when the dialog is opened or closed.
+   */
+  onOpenChangeComplete?: (open: boolean) => void;
+  /**
+   * Size of the dialog.
+   * @default "medium"
+   */
+  size?: "medium" | "small";
+}
+
+/**
+ * ..
+ * @see [📝 Documentation](https://aksel.nav.no/komponenter/core/dialog)
+ * @see 🏷️ {@link DialogProps}
+ * @example
+ * ```jsx
+ * ```
+ */
+const Dialog: React.FC<DialogProps> = (props: DialogProps) => {
+  const {
+    children,
+    defaultOpen = false,
+    open: openParam,
+    onOpenChange,
+    onOpenChangeComplete,
+    size = "medium",
+  } = props;
+
+  const [open, setOpenControlled] = useControllableState({
+    defaultValue: defaultOpen,
+    value: openParam,
+  });
+
+  const { mounted, setMounted, transitionStatus } = useTransitionStatus(open);
+
+  const popupRef = useRef<HTMLDivElement | null>(null);
+  const backdropRef = useRef<HTMLDivElement | null>(null);
+
+  const [triggerElement, setTriggerElement] = useState<HTMLElement | null>(
+    null,
+  );
+  const [popupElement, setPopupElement] = useState<HTMLElement | null>(null);
+  const [backdropElement, setBackdropElement] = useState<HTMLElement | null>(
+    null,
+  );
+
+  const defaultId = useId();
+
+  const floatingId = useMemo(() => {
+    return popupElement?.id ?? defaultId;
+  }, [defaultId, popupElement?.id]);
+
+  const [titleId, setTitleId] = useState<string>();
+
+  const [ownNestedOpenDialogs, setOwnNestedOpenDialogs] = useState(0);
+
+  const nestedDialogOpened = useEventCallback((nestedCount: number) => {
+    setOwnNestedOpenDialogs(nestedCount + 1);
+  });
+
+  const nestedDialogClosed = useEventCallback(() => {
+    setOwnNestedOpenDialogs(0);
+  });
+
+  const parentContext = useDialogContext(false);
+
+  /**
+   * Notify parent dialog about nested dialogs opening/closing.
+   * This allows us to better hide/obscure parent dialogs when nested dialogs are opened.
+   *
+   * This pattern is not good for deep nesting,
+   * but should work fine for 1-2 levels of nesting which is the most common use case here.
+   */
+  useEffect(() => {
+    if (parentContext?.nestedDialogOpened && open) {
+      parentContext.nestedDialogOpened(ownNestedOpenDialogs);
+    }
+    if (parentContext?.nestedDialogClosed && !open) {
+      parentContext.nestedDialogClosed();
+    }
+    return () => {
+      if (parentContext?.nestedDialogClosed && open) {
+        parentContext.nestedDialogClosed();
+      }
+    };
+  }, [open, parentContext, ownNestedOpenDialogs]);
```

Comment:
This could be simplified. I think this should work:

```suggestion
  useEffect(() => {
    if (!parentContext || !open) {
      return;
    }
    parentContext.nestedDialogOpened(ownNestedOpenDialogs);
    return () => {
      parentContext.nestedDialogClosed();
    };
  }, [open, parentContext, ownNestedOpenDialogs]);
```

Reply:
They are not the same, but can simplify it to:
```
useEffect(() => {
    if (
      !parentContext?.nestedDialogOpened ||
      !parentContext?.nestedDialogClosed
    ) {
      return;
    }

    if (open) {
      parentContext.nestedDialogOpened(ownNestedOpenDialogs);
    } else {
      parentContext.nestedDialogClosed();
    }

    return () => {
      open && parentContext.nestedDialogClosed();
    };
  }, [open, parentContext, ownNestedOpenDialogs]);
```

Reply:
I don't understand what you mean with "they are not the same".

1. You don't have to check if both `nestedDialogOpened` and `nestedDialogClosed` are defined because they are both required.
2. When `open` changes to `false`, the cleanup function runs and calls `nestedDialogClosed()` since it checks the "old" version of `open`. Hence the call above (in the `else`) is redundant.
3. Instead of checking `open` inside the cleanup function, you can return the cleanup function if `open` is true.